### PR TITLE
Add BandcampUser extractor

### PR DESCRIPTION
**Pull request by lyz-code**
### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [ ] New extractor
- [x] New feature

---

### Description of your *pull request* and other information

Add support for downloading list of albums from Bandcamp, to solve https://github.com/rg3/youtube-dl/issues/6231.

### CODE REVIEW

**Comment by lyz-code** [comment]
@dstftw , All checks pass now :)

File: youtube_dl/extractor/bandcamp.py [hunk 1]
```diff
@@ -415,3 +408,52 @@ def _real_extract(self, url):
             'episode_id': compat_str(video_id),
             'formats': formats
         }
+
+
+class BandcampUserIE(InfoExtractor):
+    IE_NAME = 'Bandcamp:user'
+    _VALID_URL = r'https?://(?:(?P<subdomain>[^.]+)\.)?bandcamp\.com/?$'
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: No trailing `$`, override `suitable`.

File: youtube_dl/extractor/bandcamp.py [hunk 2]
```diff
@@ -415,3 +408,52 @@ def _real_extract(self, url):
...
+        webpage = self._download_webpage(url, uploader)
+
+        album_elements = re.findall(r'<a href="/album/(.+)">', webpage)
+
+        if not album_elements:
+            raise ExtractorError('The page doesn\'t contain any albums')
+
+        entries = [
+            self.url_result(
+                compat_urlparse.urljoin(url, 'album/{}'.format(album_id)),
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: `{}` won't work in python 2.6.

File: youtube_dl/extractor/extractors.py [hunk 3]
```diff
@@ -1405,7 +1410,7 @@
     WebOfStoriesPlaylistIE,
 )
 from .weibo import (
-    WeiboIE, 
+    WeiboIE,
```
Comment (dstftw) [inline youtube_dl/extractor/extractors.py]: Remove all unrelated changes.

File: youtube_dl/extractor/bandcamp.py [hunk 4]
```diff
@@ -415,3 +408,52 @@ def _real_extract(self, url):
...
+        'playlist_count': 1,
+    }]
+
+    def _real_extract(self, url):
+        mobj = re.match(self._VALID_URL, url)
+        uploader = mobj.group('subdomain')
+
+        webpage = self._download_webpage(url, uploader)
+
+        album_elements = re.findall(r'<a href="/album/(.+)">', webpage)
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: Capture with /album. Capture non greedy.

File: youtube_dl/extractor/bandcamp.py [hunk 5]
```diff
@@ -415,3 +408,52 @@ def _real_extract(self, url):
...
+        'info_dict': {
+            'id': 'dotscale',
+            'title': 'Albums of dotscale',
+        },
+        'playlist_count': 1,
+    }]
+
+    def _real_extract(self, url):
+        mobj = re.match(self._VALID_URL, url)
+        uploader = mobj.group('subdomain')
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: `_match_id`.

File: youtube_dl/extractor/bandcamp.py [hunk 6]
```diff
@@ -415,3 +408,52 @@ def _real_extract(self, url):
...
+
+        webpage = self._download_webpage(url, uploader)
+
+        album_elements = re.findall(r'<a href="/album/(.+)">', webpage)
+
+        if not album_elements:
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: Remove.

File: youtube_dl/extractor/bandcamp.py (hunk 1 above)
Comment (lyz-code) [reply youtube_dl/extractor/bandcamp.py]: Fixed in 02d01a12fb9f4ba165d6f295251bf231268f7277 and c8aa5c673701ef6b62151e0664f41665d2c1b48f

File: youtube_dl/extractor/bandcamp.py (hunk 2 above)
Comment (lyz-code) [reply youtube_dl/extractor/bandcamp.py]: Fixed in 7eb84fd10528f3229052f896fe50fca3db101451

File: youtube_dl/extractor/extractors.py (hunk 3 above)
Comment (lyz-code) [reply youtube_dl/extractor/extractors.py]: Fixed in c17500b33a70152a9598000ba5d47cb0cd0b3967

File: youtube_dl/extractor/bandcamp.py (hunk 4 above)
Comment (lyz-code) [reply youtube_dl/extractor/bandcamp.py]: I can't capture with just /album as there are other instances that match which I don't want.

Nevertheless the non greedy capture is done in 45f47dbaa0a057dbd1daef4aba517dc947ce6853

File: youtube_dl/extractor/bandcamp.py (hunk 5 above)
Comment (lyz-code) [reply youtube_dl/extractor/bandcamp.py]: Fixed at b3b018e82775c8fd97090a85e1c1ed0083e597c5

File: youtube_dl/extractor/bandcamp.py (hunk 6 above)
Comment (lyz-code) [reply youtube_dl/extractor/bandcamp.py]: @dstftw  Fixed in 31272f0724e363253d9d1fb96f25b0bd3f94606e

Thank you for your suggestions :)

File: youtube_dl/extractor/bandcamp.py (hunk 4 above)
Comment (dstftw) [reply youtube_dl/extractor/bandcamp.py]: This does not make any difference since you prepend album anyway.

Comment (lyz-code) [reply youtube_dl/extractor/bandcamp.py]: Sorry @dstftw I wasn't understanding what you meant. Now is fixed e8f95092a

**Comment by lyz-code** [comment]
@dstftw is there anything else I need to do to make it mergeable?

**Comment by dstftw** [comment]
Read review comments and fix everything mentioned.

**Comment by lyz-code** [comment]
> Read review comments and fix everything mentioned.

Good day @dstftw, I had already answered each comment you made, they are all pending for approval

Thanks!

**Comment by dstftw** [comment]
Nope. Re-read them.

File: youtube_dl/extractor/bandcamp.py (hunk 2 above)
Comment (lyz-code) [reply youtube_dl/extractor/bandcamp.py]: More changes on this comment on ed4dd6a

**Comment by lyz-code** [comment]
> Nope. Re-read them.

After re-reading them all, I noticed there were some `{}` left, they should be fixed in the last commit.

If you feel there is something else I have to change, I'll be glad if you could tell me more specifically what it is.

I'm doing my best and it's being a little bit difficult with the short answers :sa: 

Thanks for your time

**Comment by lyz-code** [comment]
@dstftw I guess you are busy with the rest of PR. can I do anything to speed up the review process of the PR?

File: youtube_dl/extractor/bandcamp.py [hunk 7]
```diff
@@ -415,3 +408,55 @@ def _real_extract(self, url):
...
+                if BandcampAlbumIE.suitable(url) or BandcampIE.suitable(url) or
+                BandcampWeeklyIE.suitable(url)
+                else super(BandcampUserIE, cls).suitable(url))
+
+    def _real_extract(self, url):
+        uploader = self._match_id(url)
+
+        webpage = self._download_webpage(url, uploader)
+
+        album_elements = re.findall(r'<a href="/(album/.[^"]+)">', webpage)
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: Dot is pointless here.

File: youtube_dl/extractor/bandcamp.py [hunk 8]
```diff
@@ -22,6 +22,7 @@
     unified_strdate,
     unified_timestamp,
     url_or_none,
+    RegexNotFoundError,
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: flake8.

File: youtube_dl/extractor/bandcamp.py [hunk 9]
```diff
@@ -415,3 +408,55 @@ def _real_extract(self, url):
             'episode_id': compat_str(video_id),
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: 293, or is pointless.

File: youtube_dl/extractor/bandcamp.py [hunk 10]
```diff
@@ -415,3 +408,55 @@ def _real_extract(self, url):
...
             'formats': formats
         }
+
+
+class BandcampUserIE(InfoExtractor):
+    IE_NAME = 'Bandcamp:user'
+    _VALID_URL = r'https?://(?:(?P<id>[^.]+)\.)?bandcamp\.com/?'
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: `/?` does not make any sense.

File: youtube_dl/extractor/bandcamp.py [hunk 11]
```diff
@@ -415,3 +408,55 @@ def _real_extract(self, url):
...
+        album_elements = re.findall(r'<a href="/(album/.[^"]+)">', webpage)
+
+        entries = [
+            self.url_result(
+                compat_urlparse.urljoin(url, album_id),
+                ie=BandcampAlbumIE.ie_key(),
+                video_id='%s-%s' % (uploader, album_id),
+                video_title=album_id,
+            )
+            for album_id in album_elements
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: It's not an album id.

File: youtube_dl/extractor/bandcamp.py [hunk 12]
```diff
@@ -415,3 +408,55 @@ def _real_extract(self, url):
...
+
+        webpage = self._download_webpage(url, uploader)
+
+        album_elements = re.findall(r'<a href="/(album/.[^"]+)">', webpage)
+
+        entries = [
+            self.url_result(
+                compat_urlparse.urljoin(url, album_id),
+                ie=BandcampAlbumIE.ie_key(),
+                video_id='%s-%s' % (uploader, album_id),
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: Not a video id.

File: youtube_dl/extractor/bandcamp.py [hunk 13]
```diff
@@ -415,3 +408,55 @@ def _real_extract(self, url):
...
+                video_title=album_id,
+            )
+            for album_id in album_elements
+        ]
+
+        return {
+            '_type': 'playlist',
+            'id': uploader,
+            'title': 'Albums of %s' % (uploader),
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: Parentheses are pointless.

File: youtube_dl/extractor/bandcamp.py (hunk 7 above)
Comment (lyz-code) [reply youtube_dl/extractor/bandcamp.py]: Fixed

File: youtube_dl/extractor/bandcamp.py (hunk 8 above)
Comment (lyz-code) [reply youtube_dl/extractor/bandcamp.py]: Fixed

File: youtube_dl/extractor/bandcamp.py (hunk 9 above)
Comment (lyz-code) [reply youtube_dl/extractor/bandcamp.py]: Fixed

File: youtube_dl/extractor/bandcamp.py (hunk 10 above)
Comment (lyz-code) [reply youtube_dl/extractor/bandcamp.py]: Fixed

File: youtube_dl/extractor/bandcamp.py (hunk 11 above)
Comment (lyz-code) [reply youtube_dl/extractor/bandcamp.py]: Fixed

File: youtube_dl/extractor/bandcamp.py (hunk 12 above)
Comment (lyz-code) [reply youtube_dl/extractor/bandcamp.py]: I know, but I have no other way of specifying some kind of id from the information available in the user webpage.

I've removed the field.

File: youtube_dl/extractor/bandcamp.py (hunk 13 above)
Comment (lyz-code) [reply youtube_dl/extractor/bandcamp.py]: Fixed

**Comment by lyz-code** [comment]
Hi @dstftw , I'm bumping up the PR.

I've found out that there are two kinds of user webpages under the same url.

The first one has a json with all the data (including the album or track id), the other lacks that information, so I have deleted it from the entries information.

**Comment by l29ah** [comment]
Sup?

**Comment by lyz-code** [comment]
Not much @l29ah , still waiting for the reviews...

**Comment by lyz-code** [comment]
@dstftw updated the PR to match the changes on the web.

**Comment by Otiel** [comment]
@dstftw Is there still issues preventing these changes to me merged?

**Comment by lyz-code** [comment]
I've recreated the branch after the youtube-dl comeback, please @dstftw review it as you'll find that the diff is much cleaner now.

I removed the dotscale test from the album extractor because it's going to be managed by the User extractor, and will pass the url https://dotscale.bandcamp.com/album/loom to the album extractor.

Thanks
